[TG-2459] Only add counter examples when index set is exhausted

romainbrenguier opened the pull request:
Adding counter examples during the solver procedure make it very
sensitive to small changes: if the counter examples are different
between two executions then the lemmas will be different, which make the
execution diverge.
It is best to avoid them when possible, that is as long as the index set
is not exhausted.

The issue of the `indexOf` test taking a long time should disappear with this PR.

romainbrenguier committed:
Correct bounds in instantiate method

This was not checking that the instantiate variable does not exceed the
upper bound.

romainbrenguier committed:
Fix string constraints

For the PASS algorithm, the premise should be of a form that can be
analyzed by the index set construction.
For now the index set only works with the explicit bounds of the string
constraints, so the premise should not be used for now.

romainbrenguier committed:
Only add counter examples when index set exhausted

Adding counter examples during the solver procedure make it very
sensitive to small changes: if the counter examples are different
between two executions then the lemmas will be different, which make the
execution diverge.
It is best to avoid them when possible, that is as long as the index set
is not exhausted.

thk123 commented:
I can't speak to the validity of this change, but a TG pointer bump probably makes sense here.

peterschrammel commented:
@romainbrenguier, besides predictability improvements, is there a performance impact?

allredj commented:
test-gen pointer is here: https://github.com/diffblue/test-gen/pull/1581

allredj reviewed (approved):
LGTM. I was never convinced adding the counter examples was a good thing. However keeping them as a last resort for `not_contains` constraints seems reasonable. @peterschrammel we don't have a performance analysis for anything in cbmc AFAIK. It would be good to have though, but we haven't managed to push it through as a priority yet.
Adding counter-example reduces the state-space, but increases the size of the CNF, so from a theoretical point of view it could as much speed up than slow down solving. It just depends on the example at hand. 
Hopefully not adding the counter-examples will at least reduce the "randomness" of the solving.

allredj commented:
test-gen PR is failing on string tests...

tautschnig commented:
Just commenting on performance analysis: I don't know whether you have a set of tests more representative of customer workload than what is in the `regression` folder. If you have a more representative test suite, use that. If you haven't, at least time the overall runs of the regression test suite before and after this PR and report those numbers. I'd go as far as including that information in the commit message.

allredj commented:
@tautschnig yes that's a good idea and that's what we generally look at. But the figures are not very reliable because Travis has variable speed.

tautschnig commented:
> But the figures are not very reliable because Travis has variable speed.

Oh, no, never count on Travis. Do this on your own box, where you have full awareness of the load, available memory, etc. Also, re-run two or three times to get an idea of jitter.

romainbrenguier committed:
Correct find_index function for `if`-expressions

find_index would give up if the array in the index expression does not
correspond exactly to the given `str`, however it could be an expression
that contains str. For instance `(cond?str1:str2)[i+1]` is equivalent to
`cond?(str1[i+1]):(str2[i+1])`, so in both cases `find_index(e, str1, i)`
should return `i+1`.

romainbrenguier commented:
From my experiments we don't change the performances a lot (for instance see this graph for string smoke tests) except for indexOf (and a couple of similar ones) where we go from  204s to 0.06s

![perf_out](https://user-images.githubusercontent.com/11854194/37350052-c1e36c7c-26cf-11e8-9247-41d938054ede.png)

romainbrenguier committed:
Remove "Adding counter-examples" message

This can be confusing now that we don't necessarily add the counter-examples.

allredj commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -415,12 +414,17 @@ exprt string_constraint_generatort::add_axioms_for_char_set(
+  const symbol_exprt q2 = fresh_univ_index("QA_char_set2", position.type());
+  const plus_exprt lower_bound(position, from_integer(1, position.type()));
+  constraints.push_back(string_constraintt(
+    q2, lower_bound, res.length(), a3_body));
```
I find it easier to follow when constraints are named (a1, a2,...) as in the other constraint generation functions.

romainbrenguier committed:
Correct initial index set computation

The initial_index_set function was not taking into account if
expressions.
We now recursively add the components of the if expression.
The function was also performing an unecessary look-up for a quantified
variable.

allredj commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -415,12 +414,17 @@ exprt string_constraint_generatort::add_axioms_for_char_set(
-      q, from_integer(0, q.type()), res.length(), a3_guard, a3_body));
+  const equal_exprt a3_body(res[q], str[q]);
+  constraints.push_back(string_constraintt(
+    q, from_integer(0, q.type()), minimum(res.length(), position), a3_body));
```
`constraints.push_back(string_constraintt(q, minimum(res.length(), position), a3_body));`?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2125,19 +2125,20 @@ static exprt instantiate(
+        binary_relation_exprt(axiom.univ_var(), ID_ge, axiom.lower_bound()),
+        binary_relation_exprt(axiom.univ_var(), ID_lt, axiom.upper_bound())),
+      axiom.premise()),
+    axiom.body());
```
```
implies_exprt instance(
  and_exprt(
    binary_relation_exprt(axiom.univ_var(), ID_ge, axiom.lower_bound()),
    binary_relation_exprt(axiom.univ_var(), ID_lt, axiom.upper_bound()),
    axiom.premise()),
  axiom.body());
```

allredj commented on src/solvers/refinement/string_constraint_generator_transformation.cpp:
```diff
@@ -45,24 +45,23 @@ exprt string_constraint_generatort::add_axioms_for_set_length(
     idx,
-    res.length(),
-    s1.axiom_for_length_gt(idx),
+    if_exprt(binary_relation_exprt(s1.length(), ID_le, k), s1.length(), k),
```
use `minimum`?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1962,53 +1962,64 @@ static void add_to_index_set(
   }
 }
 
+/// Given an expression of the form `forall q<u. s[i]` initialize the index
```
That's a bit confusing actually. `s[i]` is not a boolean expression, so what does it mean to quantify over it?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1962,53 +1962,64 @@ static void add_to_index_set(
+  const exprt &qvar,
+  const exprt &upper_bound,
+  const exprt &s,
+  exprt i)
```
Why is `i` not `const`?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1962,53 +1962,64 @@ static void add_to_index_set(
+    initial_index_set(index_set, ns, qvar, upper_bound, ite->false_case(), i);
+    return;
+  }
+  const minus_exprt u_minus_1(upper_bound, from_integer(1, upper_bound.type()));
```
Is that the symbol case? It's not clear. Are they any assertions we can put in?

allredj commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -1962,53 +1962,64 @@ static void add_to_index_set(
+  if(s.id() == ID_array)
+  {
+    for(std::size_t j = 0; j < s.operands().size(); ++j)
+      add_to_index_set(index_set, ns, s, from_integer(j, i.type()));
```
That seems strange to me. Why do we need to add all the indices in the array case? I thought the required indices would be added later on (and not in the initial step).

romainbrenguier replied:
In an array, each character can be different, so universal formulas mentioning the array must be instantiated for each index.
For instance if `s1` is "bar" and a constraints is `forall i<3. s1[i] = s2[i]` you need `s2[0]='b'` `s2[1]='a'` and `s2[2]='r'` to get a correct model.

allredj reviewed (commented):
Looks good! Just a few minor comments.

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -2091,23 +2091,26 @@ static void update_index_set(
+/// \return an index expression in `expr` on `str` containing `qvar`. Returns
+///   an empty optional when `expr` does not contain `str`.
+static optionalt<exprt>
+find_index(const exprt &expr, const exprt &str, const symbol_exprt &qvar)
```
Not sure how this is used, but I'm wondering whether a normalisation of the if expressions would help. Btw, what would the user of this function do if the expression contains two different indices, e.g. `cond?(str1[i]):(str1[i+1])`? If this cannot happen then probably a data_invariant would be justified.

romainbrenguier replied:
Maybe an normalisation of if expression would have helped, but I'm not sure this would be preserved by simply_expr for instance, so I prefer to make function able to handle that case for now.
I added an invariant for checking there are no two such indexes, one drawback is that this does a search whose result we don't really care about, but this should be pretty cheap.

peterschrammel reviewed (approved):
LGTM once CI passes
